design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,117 @@
+* **Kubevirt**: Provide API to create a VM with minimal volume information and receive a VM with best suited populated PVC.
+* **Kubevirt**: Provide virtctl command to generate and create PVCs.
+* **SSP**: Manage setting up VolumeCloneSources for commonly supported operating systems updated by DataImportCrons.
+* **Common Templates**: Transition to use VolumeClaimTemplates in VM specs and use CDI's VolumeCloneSource as the source in the template.
+* **OCP UI**: Provide API to create a CDI Volume populator CR that can be used by the volumeClaim. In case of golden image source need to ensure VolumeCloneSource is linked properly to a PVC in order to determine whether the boot image for a common template is available.
+
+# Design
+
+A new field will be added to the VM object called **VolumeClaimTemplates**. This field will provide the user a way to define and create a PVC owned by the VM while providing partial/full PVC spec embedded in the VM spec. In case of partial spec we will make sure the preferred VolumeMode and AccessModes combination according to the StorageProfile associated with provided/default StorageClass are used.  
+This field should replace the DataVolumeTemplates field. Instead of providing a template for a DataVolume which eventually creates one, we will provide a PVC template and create it. DataVolumes and DataVolumeTemplates should be deprecated eventually.  
+A new API will be created in the CDI API library which will receive a partial PVC template and using existing CDI logic will return a complete PVC spec.  
+The provided PVC template in the VolumeClaimTemplate should have a source that can be handled by the new existing CDI volume populators API (DataSourceRef with CDI CR source) or a standalone k8s way (user costume populator, empty disk etc..)
+
+A virtctl command called `create-pvc` can be added. Given minimal flags such as name, size and source this command can generate and create a PVC. The command will use the CDI API similarly as will be done in the VM controller.
+
+## API Examples
+### VM with VolumeClaimTemplates, VolumeImportSource as volume source
+
+This example will create a VolumeImportSource along side with a VM. Once the VM is created we will generate a PVC spec with the optimal VolumeMode and AccessModes. This PVC will be created and populated with the provided image in the VolumeImportSource url.
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportSource
+metadata:
+  name: my-import-source
+spec:
+  source:
+      http:
+         url: "https://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img"
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: cirros
+spec:
+  running: true
+  volumeClaimTemplates:
+  - metadata:
+      name: cirros-pvc
```

Comment:
Let's talk about naming!

One mistake we made with `dataVolumeTemplates` was to not follow the naming pattern used by `StatefulSet`.  

Which is: `<stateful set name>-<volumeclaintemplate name>-<replica number>`.  

For example `app-mysqldb-0`.  

We don't care about replicas, so our names should be something like `myvm-mysqldb`.  

The nice thing about this scheme is `generateName` support.  Currently `generateName` support is broken for VMs with `dataVolumeTemplates`.  Ideally a user would be able to apply the same manifest with `generateName: myvm-` a hundred times.  But if you do that now, all VMs will refer to the same DataVolume.  Not good!

Reply:
Sure, this naming convention sounds good.
I thought about the generateName issue you can see the open question I mentioned in the doc. I asked community for a nice solution that was offered in this comment for example: https://github.com/kubevirt/community/pull/232#discussion_r1275112155
you can join the discussion.
In case of using generateName then in order to do get/list and track vm owned pvcs we will make sure that generateName prefix is always unique? (if using more the one volumeclaimtemplate for example)

Reply:
`generateName` should not be allowed on the `volumeClaimTemplate`.  That is how `StatefulSet` works.

`generateName` value does not have to be unique.  To get the pvcs owned by a vm, iterate `volumeClaimTemplates` and get PVC with name `<vm.Name>-<volumeClaimTemplates[i].Name> `

Reply:
If a VM useses `generateName`, and the volume, in the VolTpl, is called `myvol`, how would a disk reference a volume?

Reply:
It was discussed [here](https://github.com/kubevirt/community/pull/232#discussion_r1279048294).